### ISSUE TITLE
twisted.names.server lacks documentation and test coverage

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6700]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)| @wallrj reported|
|-|-|
|Trac ID|trac#6700|
|Type|defect|
|Created|2013-08-24 10:23:58Z|
|Branch|https://github.com/twisted/twisted/tree/names-server-coverage-6700|

In [#6645](https://github.com/twisted/twisted/issues/6645) I want to make a slight change to the behaviour of DNSServerFactory, but I found the existing tests and documentation very sparse.

Add tests and docstrings before making changes.

<details><summary>Searchable metadata</summary>

```
trac-id__6700 6700
type__defect defect
reporter__rwall rwall
priority__normal normal
milestone__ 
branch__branches_names_server_coverage_6700 branches/names-server-coverage-6700
branch_author__rwall rwall
status__closed closed
resolution__duplicate duplicate
component__names names
keywords__ 
time__1377339838000000 1377339838000000
changetime__1388758909000000 1388758909000000
version__None None
owner__rwall rwall

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_1"></a>|@wallrj commented|
|-|-|

(In [39729]) Branching to 'names-server-coverage-6700'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_2"></a>|@wallrj commented|
|-|-|

Ready for review in log:branches/names-server-coverage-6700
 * I originally intended to combine this with a fix for [#6645](https://github.com/twisted/twisted/issues/6645)
 * but the patch was getting rather large so thought it would be better reviewed in isolation.
 * Build Results:
   * http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/names-server-coverage-6700

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_3"></a>|@wallrj commented|
|-|-|

One thing I forgot to ask. Does the new test_server module need adding to the list of Python3 tests? test_names is not currently run on Python3 so I assumed not.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

Replying to [rwall](#note_3):
> One thing I forgot to ask. Does the new test_server module need adding to the list of Python3 tests? test_names is not currently run on Python3 so I assumed not.

It would really save a lot of headaches if all new code were py3k-ready.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

> It would really save a lot of headaches if all new code were py3k-ready.

True, although the cost may be headaches in making the new code py3k-ready. ;)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @wallrj<br>@exarkun set status to `new`|
|-|-|

Thanks.

  1. twisted/names/server.py
    1. `DNSServerFactory.cache` can also be `None` if no caches are specified.
    1. For `DNSServerFactory.clients`, when referencing an interface, I prefer to always be explicit about "implementer" or "provider".  This is the vocabulary `zope.interface` uses.  Typically the class is an implementer and instances of the class are providers (but things can be defined in other ways using other declaration apis from `zope.interface`).  Examples of this ambiguity appear elsewhere in the diff as well.
    1. I don't think the ''@type'' for `DNSServerFactory.protocol` adds much to the _@ivar_.  I even find it a bit confusing if interpreted literally: the callable should not actually return `IProtocolFactory`.  It should return an _IProtocolFactory provider_.  Also _L{callable}_ is incorrect here.  The `callable` builtin is just some random other API not related to the type of this instance attribute.
    1. If you care about Python 3 compat, then the signature of `DNSServerFactory.gotResolverResponse` is a problem (tuple unpacking in argument lists was removed from the language).  Not new in this branch but the function is being changed here.
    1. In `DNSServerFactory.gotResolverError`, _@param failure: A C{failure} ..._ - I'm not sure what useful information _C{failure}_ is conveying.  You might want to move the _L{...}_ from the _@type failure_ into the _@param_ to replace _C{failure}_.  Or you could be more abstract - eg, _The reason for the failed resolution (as reported by C{self.resolver.query})._  There are other instances of this elsewhere in the patch as well, eg _C{deferred}_ in `handleQuery`.
  1. twisted/names/test/test_server.py
    1. Maybe `NoresponseDNSServerFactory` should be `NoResponseDNSServerFactory`.
    1. A `pass` statement following a docstring is superfluous - eg as in `AllowQueryArguments` and `NoresponseDNSServerFactory.sendReply`.
    1. The API documentation for `NoresponseDNSServerFactory.allowQuery`'s `args` parameter seems somewhat silly.  It *must* be positional arguments because it is `*args`.  I don't think this is worth saying in a docstring.  Relatedly, though, it seems like it might be better to define these methods with the correct signature anyway.
    1. All new "root" classes should be new-style (regardless of Python 3 considerations) unless they are helpers for testing behavior related to classic classes (iow always subclass `object` if you're not subclassing something else).  It may also make sense for non-root classes to be forced to be new-style as well (mix `object` in if you're subclassing a classic class) but this is slightly more complicated so I'm not going to make a blanket statement about it.
    1. No "simple suites" - eg, no `class DummyAuthority: pass`.
    1. I think the coding standard says to use `assertIs` instead of `assertIdentical` now.
    1. `test_cacheOverride` doesn't actually demonstrate the first element is used as the cache (it could easily by the last element or the middle element or a random element and the test would still pass).
    1. The `assertIsInstance` in `test_buildProtocolProtocolOverride` seems redundant with the assertion made by `test_buildProtocolDefaultProtocolType`.  Also I find that test methods with only one assertion create a higher quality test suite than test methods with multiple assertions.
  1. The wrapping applied throughout seems to be too narrow.

I think this are largely minor comments.  I may have started to lose focus towards the end, though.  It might have been nice to have separate tickets for the documentation and testing parts of this to avoid having any >800 LOC reviewing to do (particularly given the repetitive nature of large parts of this patch (necessitated by the unfortunately repetitive nature of the API being documented) - reviewing big blocks of identical or nearly identical documentation is probably one of the harder things to do).

I just initiated a build.  Please check that for more issues that our tools catch.  Please submit for review again after addressing any of those and the above points.  Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_7"></a>|@wallrj set status to `assigned`|
|-|-|

Replying to [exarkun](#note_6):
> I think this are largely minor comments.  I may have started to lose focus towards the end, though.  It might have been nice to have separate tickets for the documentation and testing parts of this to avoid having any >800 LOC reviewing to do (particularly given the repetitive nature of large parts of this patch (necessitated by the unfortunately repetitive nature of the API being documented) - reviewing big blocks of identical or nearly identical documentation is probably one of the harder things to do).


I agree / sympathise.

Created [#6886](https://github.com/twisted/twisted/issues/6886) for missing api documentation and [#6887](https://github.com/twisted/twisted/issues/6887) for missing test coverage.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_8"></a>|@wallrj commented|
|-|-|

(In [41100]) Merge names-server-documentation-6886

Author: rwall
Reviewer: exarkun
Fixes: [#6886](https://github.com/twisted/twisted/issues/6886)
Refs: [#6700](https://github.com/twisted/twisted/issues/6700)

Add API documentation to twisted.names.server

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_9"></a>|@wallrj commented|
|-|-|

(In [41109]) Merge names-server-coverage-6887-2

Author: rwall
Reviewer: exarkun
Fixes: [#6887](https://github.com/twisted/twisted/issues/6887)
Refs: [#6700](https://github.com/twisted/twisted/issues/6700)

Add test coverage for twisted.names.server

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_10"></a>|@wallrj set status to `closed`|
|-|-|

This ticket was broken into two separate branches which have now been merged. Referenced above.
